src/Compilers/Core/Portable/Operations/OperationNodes.cs

```diff
@@ -701,4 +701,9 @@ private BlockOperation(ImmutableArray<IOperation> statements, SemanticModel sema
             Locals = ImmutableArray<ILocalSymbol>.Empty;
         }
     }
+
+    internal sealed partial class LockOperation
+    {
+        public (IOperation EnterLockScope, IMethodSymbol DisposeMethod)? LockTypeInfo { get; init; }
```

Comment:
> That will require duplicating all logic of Binder.TryFindLockTypeInfo. 

I do not think the duplication will be too big in size and complexity of that code is very low. I think we already duplicating some of the logic in the graph builder for other constructs. BTW, It might make sense for VB to build the same graphs as for C#, even though VB is going to report an error today.
Also, the decision whether this is a `Lock` based lock, should be made based on type's name, not on the availability of the APIs.

> Alternatively I guess ```Binder.TryFindLockTypeInfo``` could be changed to operate on the public ITypeSymbol instead.

That would be undesirable, compiler itself should never operate in terms of public interfaces.

Reply:
> BTW, It might make sense for VB to build the same graphs as for C#, even though VB is going to report an error today.

VB is reporting just a warning today and emits Monitor-based lock, so I think the control flow graph should stay the same for VB.

Reply:
> I do not think the duplication will be too big in size and complexity of that code is very low. I think we already duplicating some of the logic in the graph builder for other constructs.

There are other problems than just duplicating the method finding. We need to create instance call operation - but detecting whether the call should be virtual seems impossible from the control flow graph builder (I don't know how to access `receiver.SuppressVirtualCalls`).

Reply:
> VB is reporting just a warning today and emits Monitor-based lock, 

My bad, somehow I overlooked the fact. I think that should be an error and we should change that.

Reply:
> but detecting whether the call should be virtual seems impossible from the control flow graph builder (I don't know how to access ```receiver.SuppressVirtualCalls```).

We are never operating on a base reference, therefore "our" virtual calls should never be suppressed based on the receiver.